doc/api/n-api.md

```diff
@@ -3705,6 +3730,93 @@ NAPI_EXTERN napi_status napi_get_uv_event_loop(napi_env env,
 - `[in] env`: The environment that the API is invoked under.
 - `[out] loop`: The current libuv loop instance.
 
+<!-- it's very convenient to have all the anchors indexed -->
+<!--lint disable no-unused-definitions remark-lint-->
+## Asynchronous Thread-safe Function Calls
+JavaScript functions can normally only be called from a native addon's main
+thread. If an addon creates additional threads then N-API functions that require
+a `napi_env`, `napi_value`, or `napi_ref` must not be called from those threads.
+
+This API provides the type `napi_threadsafe_function` as well as APIs to create,
+destroy, and call objects of this type. `napi_threadsafe_function` creates a
+permanent reference to a `napi_value` that holds a JavaScript function, and
+uses `uv_async_t` from libuv to coordinate calls to the JavaScript function from
+all threads.
+
+Since `uv_async_t` is used in the implementation, the caveat whereby multiple
+invocations on the secondary thread may result in only one invocation of the
+JavaScript function also applies to `napi_threadsafe_function`.
```

Comment:
There's no reason it must be that way.  Add a mutex and a list or queue, then the `uv_async_t` becomes like an event-driven condition variable.

Reply:
@bnoordhuis @nodejs/addon-api This is indeed value that we could add and this was indeed [requested](https://github.com/nodejs/node/issues/13512#issuecomment-346845043), however, implementing this requires answers to a host of design questions, as the linked comment reveals towards the end.

So, we need to discuss how far beyond exposing libuv and Node functionality we wish to go.

If we decide to implement this queuing, then, in addition to the questions raised by the comment, we need to decide whether to drain the queue using a busy loop or an idle loop, and what limit we should place on the length of the queue. Do we expose both `lock` and `trylock` for the mutex so that users might choose whether to block the secondary thread when the queue gets full? How do we dispose of the `void*` data once the JavaScript function has been called? How close would using such a queue bring this implementation to that of a pipe, and is it worth at that point starting down this path?

TBH, I was going in the opposite direction with this PR (that is, providing less value beyond the essential, rather than more), but I haven't yet had a chance to submit the changes. Instead of a `marshal_cb` and `process_return_cb`, I was just going to go with
```C
typedef void (*napi_threadsafe_function_call_js)(napi_env env, napi_value func, void* data);
```
thereby unifying the marshalling and return value processing step. I did that because I realized that we have a lot of `if (status != napi_ok) { napi_throw_error(); }` and `if (status != napi_ok) { napi_fatal_error(); }` on the N-API side of the code, and that's perhaps not how the user would want to deal with such failures. Having a single function pointer would push this handling onto the user. Nevertheless, the function pointer would be called from withing a `node::CallbackScope` so we wouldn't have to remind the user to call the function via `napi_make_callback()`.

I suppose the deeper question is whether we want users using libuv at all. After all, the `void*` could be a `struct { uv_mutex_t mutex; /* useful and thread-safe stuff here; */ };` and then the user would be implementing their own queuing, or any other semantics.

In fact, my only reason for originally splitting the marshalling and return value processing into two steps was to ensure that the actual calling of the JavaScript function was under my control, whereby I could ensure that the call happened via `napi_make_callback()`. Yet in that incarnation I wasn't using any Node internal APIs, meaning that this whole PR could be implemented externally to Node.